game-core/src/main/java/games/strategy/engine/pbem/NodeBbForumPoster.java

```diff
@@ -0,0 +1,171 @@
+ *
+ * <p>
+ * URL format is {@code https://your.forumurl.com/api/v2/topics/<topicID>}.
+ * </p>
+ */
+@Log
+abstract class NodeBbForumPoster extends AbstractForumPoster {
+  private static final long serialVersionUID = -6128723809566917089L;
+
+  protected abstract String getForumUrl();
+
+  @Override
+  public boolean postTurnSummary(final String summary, final String title) {
+    try (CloseableHttpClient client = HttpClients.custom().disableCookieManagement().build()) {
+      final int userId = getUserId(client);
+      final String token = getToken(client, userId);
+      try {
+        post(client, token, "### " + title + "\n" + summary);
+        turnSummaryRef = "Successfully posted!";
+        return true;
+      } finally {
+        deleteToken(client, userId, token);
+      }
+    } catch (final IOException | IllegalStateException e) {
+      log.log(Level.SEVERE, "Failed to post game to forum", e);
+      turnSummaryRef = e.getMessage();
+    }
+    return false;
+  }
+
+  private void post(final CloseableHttpClient client, final String token, final String text) throws IOException {
+    final HttpPost post = new HttpPost(getForumUrl() + "/api/v2/topics/" + getTopicId());
+    addTokenHeader(post, token);
+    post.setEntity(new UrlEncodedFormEntity(
+        Collections.singletonList(new BasicNameValuePair("content",
+            text + ((m_includeSaveGame && saveGameFile != null) ? uploadSaveGame(client, token) : ""))),
+        StandardCharsets.UTF_8));
+    HttpProxy.addProxy(post);
+    try (CloseableHttpResponse response = client.execute(post)) {
+      final int code = response.getStatusLine().getStatusCode();
+      if (code != HttpURLConnection.HTTP_OK) {
+        throw new IllegalStateException("Forum responded with code " + code);
+      }
+    }
+  }
+
+  private String uploadSaveGame(final CloseableHttpClient client, final String token) throws IOException {
+    final HttpPost fileUpload = new HttpPost(getForumUrl() + "/api/v2/util/upload");
+    fileUpload.setEntity(MultipartEntityBuilder.create()
+        .addBinaryBody("files[]", saveGameFile, ContentType.APPLICATION_OCTET_STREAM, saveGameFileName)
+        .build());
+    HttpProxy.addProxy(fileUpload);
+    addTokenHeader(fileUpload, token);
+    try (CloseableHttpResponse response = client.execute(fileUpload)) {
+      final int status = response.getStatusLine().getStatusCode();
+      if (status == HttpURLConnection.HTTP_OK) {
+        final String json = EntityUtils.toString(response.getEntity());
+        return "\n[Savegame](" + new JSONArray(json).getJSONObject(0).getString("url") + ")";
+      }
+      throw new IllegalStateException("Failed to upload savegame, server returned Error Code " + status);
+    }
+  }
+
+  private void deleteToken(final CloseableHttpClient client, final int userId, final String token)
+      throws IOException {
+    final HttpDelete httpDelete = new HttpDelete(getForumUrl() + "/api/v2/users/" + userId + "/tokens/" + token);
```

Comment:
Is it necessary to call `HttpProxy#addProxy()` on this request?

If so, it might be worth it to extract a private helper that does that and `addTokenHeader()` at the same time.  Maybe even just have the method do all that as well as call `client.execute(...)` so it can all be encapsulated as part of creating the response.

**EDIT:** I see how you just moved this code from `TripleAForumPoster`, so feel free to ignore the refactoring suggestion.  However, please determine if missing the proxy configuration is a bug here.

Reply:
The proxy is added to all requests, so I guess it's required here as well.

Reply:
Since you're working in this area, would you mind fixing that in a follow-up PR?

Reply:
Maybe, if you can tell me how exactly you are planning to simplify all of those HTTP-Request creations.
Probably not in the next PR, even by barely looking at the code the next PR regarding this topic will probably already be way too huge.

Reply:
Although refactoring this code would be nice, I think simply adding the missing `HttpProxy.addProxy(httpDelete)` call would be sufficient.  So, basically a one-line change. :smile:

Reply:
Ah, thanks for clarification, I misunderstood your request 😅 
In this case I'll create a small follow-up PR containing this change.
The bigger change requires a lot more changes than I thought.
I looked again at the wiring of the classes implementing IBean and oh boy,  a lot of code is depending on it.
I think I could really use some help with the design approach.
I'm pretty sure we don't want to serialize IForumPoster and IEmailSender into every PBF/PBEM savegame, especially because of the password security issue you thankfully fixed.
That's mainly why I'm asking you: I'm not entirely sure how much should actually be sent via the PBEM/PBF savegame?

As a Side note: I recently learned a bit about the Java Serialization format and how it can be read on a binary level and for whatever reason the game notes seem to get serialized into the savegame as well. I don't think that's supposed to happen.

Reply:
Also any ideas on how to store the things that should get shipped with the savegame effectively into the map properties?

Reply:
> I think I could really use some help with the design approach.

It's been a long time since I looked at this code, but IIRC, the reason `IForumPoster` and `IEmailSender` are part of the save game is so the players don't have to keep typing in all the fields (sans username and password) during each turn.  I think whoever wrote the code originally just thought it would be simpler to reuse the existing types that are saved in the local bean cache (and wipe the credentials) rather than creating totally separate types that only contain the required information.

I think "user preferences" and "save game" are distinct-enough domains that having separate types to store PBEM/PBF information for each domain wouldn't be a design smell.  I'd rather have some code that looks kind of similar than coupling between two unrelated domains.

I'd say the best thing to do is take a shot at a first iteration, and then we can discuss it.

> Also any ideas on how to store the things that should get shipped with the savegame effectively into the map properties?

Not sure I understand the question.  By "effectively," do you mean _efficiently_?